router/batchrouter/partition_worker.go

```diff
@@ -0,0 +1,240 @@
+package batchrouter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	kitsync "github.com/rudderlabs/rudder-go-kit/sync"
+	obskit "github.com/rudderlabs/rudder-observability-kit/go/labels"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	asynccommon "github.com/rudderlabs/rudder-server/router/batchrouter/asyncdestinationmanager/common"
+	"github.com/rudderlabs/rudder-server/router/batchrouter/circuitbreaker"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	warehouseutils "github.com/rudderlabs/rudder-server/warehouse/utils"
+)
+
+type PartitionWorker struct {
+	wg      sync.WaitGroup
+	logger  logger.Logger
+	brt     *Handle
+	channel chan *JobEntry
+	cb      circuitbreaker.CircuitBreaker
+	limiter kitsync.Limiter
+
+	addJobMaxDelay time.Duration
+}
+
+type JobEntry struct {
+	job      *jobsdb.JobT
+	sourceID string
+	destID   string
+}
+
+func NewPartitionWorker(logger logger.Logger, partition string, brt *Handle, cb circuitbreaker.CircuitBreaker) *PartitionWorker {
+	maxJobsToBuffer := brt.conf.GetIntVar(100000, 1, "BatchRouter."+brt.destType+".partitionWorker.maxJobsToBuffer", "BatchRouter.partitionWorker.maxJobsToBuffer")
+
+	pw := &PartitionWorker{
+		channel: make(chan *JobEntry, maxJobsToBuffer),
+		logger:  logger.With("partition", partition),
+		brt:     brt,
+		cb:      cb,
+		limiter: kitsync.NewReloadableLimiter(
+			context.Background(), &sync.WaitGroup{}, "batchrouter_partition_worker",
```

Comment:
limiter runs goroutines internally to report stats, that's why we need to provide a valid context and waitgroup. This context needs to be canceled when we stop the worker & the worker needs to wait for the waitgroup to finish